bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,237 @@
+            newFile.createNewFile();
+            initAppendChannel(newFile);
+            if (existingLogs == null) {
+                return newFile;
+            }
+            // write existing unfinished logs to the new file
+            if (!existingLogs.isEmpty()) {
+                cleanUpFinishedLogs();
+                if (!existingLogs.isEmpty()) {
+                    for (Map.Entry<String, TransactionLogRecord> entry : existingLogs.entrySet()) {
+                        put(entry.getValue());
+                    }
+                    existingLogs.clear();
+                }
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to create recovery log file in " + recoveryLogDir);
+        }
+        return newFile;
+    }
+
+    /**
+     * Finds the latest version of the recovery log file.
+     *
+     * @return The latest version of the recovery log file.
+     */
+    private int findLatestVersion() {
+        int latestVersion = 0;
+        File directory = recoveryLogDir.toFile();
+        File[] files = directory.listFiles((dir, name) -> name.matches(baseFileName + "(\\d+)\\.log"));
+        if (files == null) {
+            return latestVersion;
+        }
+        for (File file : files) {
+            String fileName = file.getName();
+            int version = Integer.parseInt(fileName.replaceAll(
+                    baseFileName, "").replaceAll(".log", "")
+            );
+            if (version > latestVersion) {
+                latestVersion = version;
+            }
+        }
+        return latestVersion;
+    }
+
+    /**
+     * Initializes the append channel for the given file.
+     *
+     * @param file The file to initialize the append channel for.
+     */
+    private void initAppendChannel(File file) {
+        try {
+            appendChannel = FileChannel.open(file.toPath(), StandardOpenOption.APPEND);
+            FileLock lock = appendChannel.tryLock();
+            if (lock == null) {
+                stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+        }
+    }
+
+    @Override
+    public void put(TransactionLogRecord trxRecord) {
+        boolean force = !(trxRecord.getTransactionState().equals(RecoveryState.TERMINATED)); // lazy write for done record
+        writeToFile(trxRecord.getTransactionLogRecord(), force);
+        if (checkpointInterval != -1) {
+            ifNeedWriteCheckpoint();
+            numOfPutsSinceLastCheckpoint++;
+        }
+    }
+
+    public Map<String, TransactionLogRecord> getPendingLogs() {
+        Map<String, TransactionLogRecord> pendingTransactions = new HashMap<>();
+        Map<String, TransactionLogRecord> transactionLogs = readLogsFromFile(file);
+        if (transactionLogs == null) {
+            return null;
+        }
```

Comment:
IMO it's better to use `Optional` rather than returning `null` which might case NPEs